src/v/archival/ntp_archiver_service.cc

```diff
@@ -707,46 +708,59 @@ ntp_archiver::schedule_uploads(std::vector<upload_context> loop_contexts) {
             _probe.upload_lag(ctx.last_offset - ctx.start_offset);
         }
 
         co_await ss::repeat(
-          [this, &ctx, &uploads_remaining]() -> ss::future<ss::stop_iteration> {
-              if (uploads_remaining <= 0 || !upload_loop_can_continue()) {
+          [this, &ctx, uploads_remaining]() -> ss::future<ss::stop_iteration> {
+              if (*uploads_remaining <= 0 || !upload_loop_can_continue()) {
                   co_return ss::stop_iteration::yes;
               }
 
-              auto should_stop = co_await ctx.schedule_single_upload(*this);
-
-              // At the end of each context is an upload with should_stop=yes,
-              // which does not upload anything but signals the end of the
-              // uploads. Do not decrement remaining uploads for this type of
-              // return value. When should_stop=no, we have a real upload and
-              // decrement the counter.
-              if (should_stop == ss::stop_iteration::no) {
-                  uploads_remaining -= 1;
-              }
-              co_return should_stop;
-          });
-
-        vlog(
-          _rtclog.debug,
-          "scheduled {} uploads for upload kind: {}, uploads remaining: {}",
-          ctx.uploads.size(),
-          ctx.upload_kind,
-          uploads_remaining);
-
-        scheduled_uploads.insert(
-          scheduled_uploads.end(),
-          std::make_move_iterator(ctx.uploads.begin()),
-          std::make_move_iterator(ctx.uploads.end()));
+              co_return co_await ctx.schedule_single_upload(*this).then(
+                ss::coroutine::lambda([uploads_remaining](auto should_stop) {
+                    // At the end of each context is an upload with
+                    // should_stop=yes, which does not upload anything but
+                    // signals the end of the uploads. Do not decrement
+                    // remaining uploads for this type of return value. When
+                    // should_stop=no, we have a real upload and decrement the
+                    // counter.
+                    if (should_stop == ss::stop_iteration::no) {
+                        *uploads_remaining -= 1;
+                    }
+                    return should_stop;
+                }));
+          })
+          .then(ss::coroutine::lambda([this,
+                                       uploads_remaining,
+                                       &ctx,
+                                       scheduled_uploads]() {
+              vlog(
+                _rtclog.debug,
+                "scheduled {} uploads for upload kind: {}, uploads remaining: "
+                "{}",
+                ctx.uploads.size(),
+                ctx.upload_kind,
+                *uploads_remaining);
+
+              scheduled_uploads->insert(
+                scheduled_uploads->end(),
+                std::make_move_iterator(ctx.uploads.begin()),
+                std::make_move_iterator(ctx.uploads.end()));
+          }));
     }
```

Comment:
Could we avoid both of these lambdas by rewriting this as:

```
    while (true) {
        if (uploads_remaining <= 0 || !upload_loop_can_continue()) {
            break;
        }
        auto should_stop = co_await ctx.schedule_single_upload(*this);
        if (should_stop == ss::stop_iteration::yes) {
            break;
        }
        uploads_remaining -= 1;
        ss::coroutine::maybe_yield();
    }
    vlog(...)
    scheduled_uploads.insert(
        scheduled_uploads.end(),
        std::make_move_iterator(ctx.uploads.begin()),
        std::make_move_iterator(ctx.upload.end()));
}
co_return scheduled_uploads;
```

Reply:
Yes, I think this is doable. In fact, we can even skip the `maybe_yield` because the `co_await` for scheduling single upload provides a yield point and the processing after in the loop body is trivial. I will test it out locally. Thanks for the suggestion.